### PR TITLE
Fix: Properly rewrite rich text with mixed links

### LINKED ISSUE #11957: Result of `expand_db_html` for a mixture of internal / external links is malformed
### Issue Summary

Running `expand_db_html` on rich text input containing a mix of internal and external links results in malformed output, presumably as a consequence of #5875.

### Steps to Reproduce

Within `./manage.py shell` on a bakerydemo instance:

```python
from wagtail.rich_text import expand_db_html
rich_text = '<p data-block-key="9l01v"><a href="https://wagtail.org/">External link 1</a></p>'\
            '<p data-block-key="b1mdf"><a linktype="page" id="3">Internal link</a></p>'\
            '<p data-block-key="1hsrl"><a href="https://wagtail.org/">External link 2</a></p>'
expand_db_html(rich_text)
```

results in

```
'<p data-block-key="9l01v"><a href="/breads/">External link 1</a></p><p data-block-key="b1mdf"><a linktype="page" id="3">Internal link</a></p><p data-block-key="1hsrl"><a href="https://wagtail.org/">External link 2</a></p>'
```

Note that the URL `/breads/` (corresponding to page ID 3) has been substituted on external link 1 instead of the internal link as expected.

Additionally, with [the NoFollowExternalLinkHandler example from the Wagtail docs](https://docs.wagtail.org/en/stable/extending/rich_text_internals.html#FeatureRegistry.register_link_type) defined in `base/wagtail_hooks.py`:

```python
from wagtail.rich_text import LinkHandler
from django.utils.html import escape
from wagtail import hooks


class NoFollowExternalLinkHandler(LinkHandler):
    identifier = 'external'

    @classmethod
    def expand_db_attributes(cls, attrs):
        href = attrs["href"]
        return '<a href="%s" rel="nofollow">' % escape(href)


#@hooks.register('register_rich_text_features')
def register_external_link(features):
    features.register_link_type(NoFollowExternalLinkHandler)
```

we get the output

```
'<p data-block-key="9l01v"><a href="https://wagtail.org/" rel="nofollow">External link 1</a></p><p data-block-key="b1mdf"><a linktype="pa<a href="/breads/">p><p data-block-key="1hsrl"><a href="https://wagtail.org/" rel="nofollow">External link 2</a></p>'
```

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

-   Python version: Run `python --version`.
-   Django version: Look in your requirements.txt, or run `pip show django | grep Version`.
-   Wagtail version: Look at the bottom of the Settings menu in the Wagtail admin, or run `pip show wagtail | grep Version:`.
-   Browser version: You can use https://www.whatsmybrowser.org/ to find this out.

### Working on this

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

### PR DESCRIPTION
**Pull request by chosak**
The bulk rich text link rewriter logic introduced in  #5875 doesn't work properly if the rich text contains multiple link types where one of those types doesn't need to be rewritten (for example external links like `<a href="https://wagtail.org/">`).

This commit fixes that bug by handling these cases, and adds additional unit tests to cover the failure cases.

Fixes #11957.

### CODE REVIEW

**Comment by gasman** [comment]
Thanks @chosak! I see the second issue mentioned on #11957, where a custom LinkHandler is in use, is still present - I think this might be a separate bug, since it involves the replacement tag being written in the wrong place entirely, rather than just URLs being switched around.

**Comment by chosak** [comment]
Thanks @gasman! I've just pushed a fix and test for that issue as well.

**Comment by gasman** [comment]
Thanks @chosak! I'm happy with the updated logic, although still concerned that the "zipping tuples of dicts" approach is rather hard to follow, and probably largely responsible for these bugs sneaking in in the first place...

I've now had a go at refactoring it by introducing a `TagMatch` object to encapsulate all the information for an individual tag - please let me know what you think.

**Comment by chosak** [comment]
@gasman works for me! This is a nice way to make the logic more straightforward, thank you.

**Comment by gasman** [comment]
Thanks @chosak! Merged in bf3f87b75910495471e84e976fff0af7f6a74df5 (main) / 729acab069661e83bd6ee429ffc73bdd59068ccb (stable/6.1.x).
